File: workspace/bible/processor/src/usx/list.ts
/*
 * Copyright 2024 Sowers, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { ListStyle, ListStyles } from './styles'
import { Attributes, StyleFactoryFilter, UsxContext, UsxItem, UsxItemContainer, UsxItemFactory } from './item'
import { Reference, ReferenceFactory } from './reference'
import { Footnote, FootnoteFactory } from './footnote'
import { Char, CharFactory } from './char'
import { Milestone, MilestoneFactory } from './milestone'
import { Figure, FigureFactory } from './figure'
import { Verse } from './verse'
import { Break, BreakFactory } from './break'
import { Text, TextFactory } from './text'
import { CrossReference, CrossReferenceFactory } from './cross_reference'
import { ListChar, ListCharFactory } from './list_char'
import { VerseStartFactory } from './verse_start'
import { VerseEndFactory } from './verse_end'

type ListType = Reference | Footnote | CrossReference | Char | ListChar | Milestone | Figure | Verse | Break | Text

export class List extends UsxItemContainer<ListType> {

  style: ListStyle
  vid: string | null

  constructor(context: UsxContext, parent: UsxItem | null, attributes: Attributes) {
    super(context, parent, attributes)
    this.style = attributes.STYLE.toString() as ListStyle
    this.vid = attributes.VID?.toString() || null
  }

  get htmlClass(): string {
    return this.style
  }
}

export class ListFactory extends UsxItemFactory<List> {

  static readonly instance = new ListFactory()

  private constructor() {
    super('para', new StyleFactoryFilter(ListStyles))
  }

  protected onInitialize() {
    this.register(ReferenceFactory.instance)
    this.register(FootnoteFactory.instance)
    this.register(CrossReferenceFactory.instance)
    this.register(CharFactory.instance)
    this.register(ListCharFactory.instance)
    this.register(MilestoneFactory.instance)
    this.register(FigureFactory.instance)
    this.register(VerseStartFactory.instance)
    this.register(VerseEndFactory.instance)
    this.register(BreakFactory.instance)
    this.register(TextFactory.instance)
  }

  create(context: UsxContext, parent: UsxItem | null, attributes: Attributes): List {
    return new List(context, parent, attributes)
  }
}

File: workspace/bible/processor/src/usx/list.rs
use crate::context::UsxContext;
use crate::singleton::{FactoryHandle, FactorySingleton};
use crate::usx::char::CharFactory;
use crate::usx::cross_reference::CrossReferenceFactory;
use crate::usx::factory::{BaseFactory, IUsxItemFactory, UsxItemFactory};
use crate::usx::figure::FigureFactory;
use crate::usx::footnote::FootnoteFactory;
use crate::usx::item::{ItemFactoryFilter, StyleFactoryFilter, UsxItem, UsxItemContainer};
use crate::usx::list_char::ListCharFactory;
use crate::usx::milestone::MilestoneFactory;
use crate::usx::optbreak::BreakFactory;
use crate::usx::reference::ReferenceFactory;
use crate::usx::styles::ListStyle;
use crate::usx::text::TextFactory;
use crate::usx::verse_end::VerseEndFactory;
use crate::usx::verse_start::VerseStartFactory;
use std::collections::HashMap;
use std::sync::{Arc, Mutex};

pub struct List {
    pub style: ListStyle,
    pub vid: Option<String>,
    pub container: UsxItemContainer,
}

impl List {
    pub fn create(
        context: &mut UsxContext,
        _parent: &Option<Arc<Mutex<UsxItem>>>,
        attributes: &HashMap<String, String>,
    ) -> Arc<Mutex<UsxItem>> {
        let container = UsxItemContainer::new(context);
        let item = Arc::new(Mutex::new(UsxItem::List(List {
            style: attributes.get("style").unwrap().as_str().into(),
            vid: attributes.get("vid").cloned(),
            container,
        })));
        item
    }
}

static mut INSTANCE: FactorySingleton = FactorySingleton::new();

fn factory() -> Arc<UsxItemFactory> {
    Arc::new(UsxItemFactory::List(ListFactory {
        base: BaseFactory::new(
            "para",
            Some(ItemFactoryFilter::Style(StyleFactoryFilter::new(
                ListStyle::to_str_name(),
            ))),
        ),
    }))
}

#[derive(Debug)]
pub struct ListFactory {
    base: BaseFactory,
}

impl ListFactory {
    pub fn get() -> &'static FactoryHandle {
        unsafe {
            INSTANCE.initialize(factory);
            INSTANCE.get()
        }
    }
}

impl IUsxItemFactory for ListFactory {
    fn base_factory(&self) -> &BaseFactory {
        &self.base
    }

    fn base_factory_mut(&mut self) -> &mut BaseFactory {
        &mut self.base
    }

    fn on_initialize(&mut self) {
        self.register(ReferenceFactory::get());
        self.register(FootnoteFactory::get());
        self.register(CrossReferenceFactory::get());
        self.register(CharFactory::get());
        self.register(ListCharFactory::get());
        self.register(MilestoneFactory::get());
        self.register(FigureFactory::get());
        self.register(VerseStartFactory::get());
        self.register(VerseEndFactory::get());
        self.register(BreakFactory::get());
        self.register(TextFactory::get());
    }

    fn create(
        &self,
        context: &mut UsxContext,
        parent: &Option<Arc<Mutex<UsxItem>>>,
        attributes: &HashMap<String, String>,
    ) -> Arc<Mutex<UsxItem>> {
        List::create(context, parent, attributes)
    }
}
